README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[English](README.md) | [日本語](README_JP.md)

# NumComma

<!-- badges: start -->

<!-- badges: end -->

`{NumComma}` can convert between comma-separated and non-comma-separated
numbers every three digits on R.

It also has the `*_str` function for handling comma-separated and
non-comma-separated numbers in a string.

## Installation

You can install it in

``` r
remotes::install_github("indenkun/NumComma")
```

## Example

``` r
library(NumComma)
```

`{NumComma}` has two functions: `NumComma_rm` for converting
comma-separated numbers into uncomma-separated numbers every three
digits, and `NumComma_add` for converting numbers into comma-separated
numbers (strings) every three digits. And there are extensions to these
functions, `NumComma_rm_str` and `NumComma_add_str`, which perform the
conversion for each digit in a string.

### `NumComma_rm` and `NumComma_rm_str`

`NumComma_rm` can convert comma-separated numbers (strings) into numbers
without commas every three digits.

``` r
NumComma_rm("2,020")
#> [1] "2020"
```

If the number does not contain a comma, it returns the same number.

``` r
NumComma_rm("2020")
#> [1] "2020"
```

Numbers containing numbers after the decimal point can also be
converted.

``` r
NumComma_rm("1,234.56")
#> [1] "1234.56"
```

Numbers containing numbers after the decimal point can also be
converted.

``` r
x <- c("2,020", "2020", "1,234.56")
NumComma_rm(x)
#> [1] "2020"    "2020"    "1234.56"
```

It is also possible to return the value as a number instead of a string
by using `convert` with `TRUE`.

``` r
x <- c("2,020", "2020", "1,234.56")
NumComma_rm(x, convert = TRUE)
#> [1] 2020.00 2020.00 1234.56
```

If you enter a number with a comma or a non-comma value for every three
digits, it returns the value as it was entered with a warning message.

``` r
NumComma_rm("1,23,456")
#> Warning in .f(.x[[1L]], .y[[1L]], ...): this format cannot be converted.
#> [1] "1,23,456"
NumComma_rm("NumComma_rm")
#> Warning in .f(.x[[1L]], .y[[1L]], ...): Strings containing characters other than
#> numbers and commas cannot be converted.
#> [1] "NumComma_rm"
```

The warning text can be disabled by specifying `FALSE` in
`Warnig.message`.

``` r
NumComma_rm(c("1,23,456", "NumComma_rm"), warnig.message = FALSE)
#> [1] "1,23,456"    "NumComma_rm"
```

`NumComma_rm_str` is a function for converting a comma-separated number
to an uncomma-separated number for every three digits in a string.

``` r
NumComma_rm_str("The total population of Japan is 124,271,318 as of January 1, 2020.")
#> [1] "The total population of Japan is 124271318 as of January 1, 2020."
```

### `NumComma_add` and `NumComma_add_str`

NumComma\_add\` is a function for converting numbers into
comma-separated numbers (strings) of three digits.

It can accept a number or a string that represents a number.

``` r
NumComma_add(123456)
#> [1] "123,456"
NumComma_add("123456")
#> [1] "123,456"
```

If you enter consecutive numbers starting from 0 as a string it will
return the value as is.

``` r
NumComma_add("0001")
#> [1] "0001"
```

If you enter a string that is not a number string (including numbers
containing commas), it returns the value you entered with a warning
message.

``` r
NumComma_add(c("NumComma_add", "1,234"))
#> Warning in .f(.x[[i]], ...): Strings containing characters other than numbers
#> cannot be converted.

#> Warning in .f(.x[[i]], ...): Strings containing characters other than numbers
#> cannot be converted.
#> [1] "NumComma_add" "1,234"
```

The warning text can be disabled by specifying `FALSE` in
`Warnig.message`.

``` r
NumComma_add(c("NumComma_add", "1,234"), warnig.message = FALSE)
#> [1] "NumComma_add" "1,234"
```

You can also use `digit` to convert only numbers with more than the
number of digits (including decimal places) specified by `digit` into
comma-separated numbers with every three digits, or `small.num` to
convert only numbers with more than the number of digits specified by
`small.num` into comma-separated numbers with every three digits.

``` r
NumComma_add(c(123456, 2020), digit = 4)
#> [1] "123,456" "2020"
NumComma_add(c(123456, 2020), small.num = 3000)
#> [1] "123,456" "2020"
```

`NumComma_add_str` is a function for converting numbers in a string to
comma-separated numbers.

``` r
NumComma_add_str("The total population of Japan is 124271318 people as of January 1, 2020.")
#> [1] "The total population of Japan is 124,271,318 people as of January 1, 2,020."
```

The `NumComma_add_str`, like the `NumComma_add`, can control the number
to be converted by arguments.

``` r
NumComma_add_str("The total population of Japan is 124271318 people as of January 1, 2020.", digit = 4)
#> [1] "The total population of Japan is 124,271,318 people as of January 1, 2020."
NumComma_add_str("The total population of Japan is 124271318 people as of January 1, 2020.", small.num = 3000)
#> [1] "The total population of Japan is 124,271,318 people as of January 1, 2020."
```

## License

MIT.

## Imports

  - `{purrr}`
  - `{stringr}`
  - `{stats}`
  - `{dplyr}`


NEWS.md
# NumComma 0.0.0.9000

* Optimized the code.
* Added a word list to support spell checking.
* Reviewed namespaces.

# NumComma 0.0.0.1

* Initial Release to GitHub.


tests/testthat.R
library(testthat)
library(NumComma)

test_check("NumComma")


tests/testthat/test-NumComma.R
test_that("Convert to comma-separated numbers", {
  expect_equal(NumComma_add(2020), "2,020")
  expect_equal(NumComma_add(c(2019, 2020, 2021)),
               c("2,019","2,020", "2,021"))
  expect_equal(NumComma_add(c(123456, 1234), digit = 4),
               c("123,456", "1234"))
  expect_equal(NumComma_add(c(123456, 1234), small.num = 2000),
               c("123,456", "1234"))
  expect_equal(NumComma_add(1234.56),
               "1,234.56")
  expect_equal(NumComma_add(1234.56, digit = 4),
               "1,234.56")
  expect_equal(NumComma_add(1234.56, small.num = 2000),
               "1234.56")
  expect_warning(NumComma_add("japan"))
  expect_warning(NumComma_add("1,23,456"))
  expect_equal(NumComma_add("japan", warnig.message = FALSE), "japan")
  expect_equal(NumComma_add("0007"), "0007")
})

test_that("Convert to comma-separeted numbers in strings", {
  expect_equal(NumComma_add_str("This year is 2020 and next year is 2021."),
               "This year is 2,020 and next year is 2,021.")
  expect_equal(NumComma_add_str("This is 123456, but that's 2020.", digit = 4),
               "This is 123,456, but that's 2020.")
  expect_equal(NumComma_add_str("This is 123456, but that's 2020.", small.num = 3000),
               "This is 123,456, but that's 2020.")
})

test_that("Convert to non comma-separeted numbers", {
  expect_equal(NumComma_rm("2,020"), "2020")
  expect_equal(NumComma_rm("2020"), "2020")
  expect_equal(NumComma_rm("2,020", convert = TRUE), 2020)
  expect_equal(NumComma_rm("2020", convert = TRUE), 2020)
  expect_equal(NumComma_rm(2020, convert = TRUE), 2020)
  expect_equal(NumComma_rm(c("2,020", "2,021")), c("2020", "2021"))
  expect_warning(NumComma_rm("japan"))
  expect_warning(NumComma_rm("1,23,456"))
  expect_equal(NumComma_rm("japan", warnig.message = FALSE), "japan")
  expect_equal(NumComma_rm("1,23,456", warnig.message = FALSE), "1,23,456")
  expect_equal(NumComma_rm("1,234.56"), "1234.56")
  expect_equal(NumComma_rm(c("1,234", "1,234.56", "1,23", "japan"), warnig.message = FALSE),
               c("1234", "1234.56", "1,23", "japan"))
  expect_equal(NumComma_rm(c("1,234", "1,234.56", "1,23", "japan"), convert = TRUE, warnig.message = FALSE),
               c(1234, 1234.56, NA, NA))
  expect_equal(NumComma_rm("0007"), "0007")
})

test_that("Convert to non comma-separeted numbers in strings", {
  expect_equal(NumComma_rm_str("This year is 2,020 and next year is 2,021."),
               "This year is 2020 and next year is 2021.")
  expect_equal(NumComma_rm_str("This year is 2,0,2,0 and next year is 2,021."),
               "This year is 2,0,2,0 and next year is 2021.")
})


R/NumComma.R
#' Numbers and commas
#' @description
#' `NumComma_rm`, which takes a comma-separated number every three digits as a
#' string and converts it into a number without a comma. `NumComma_rm_str` is a
#' function to convert comma-separated numbers in a string into comma-separated
#' numbers without commas for every three digits while keeping the rest of the
#' string. While `NumComma_rm` accepts only positive numbers, `NumComma_rm_str`
#' accepts comma-separated numbers with three digits.
#' `NumComma_add` is a function that converts a number not containing a comma to
#'  a comma-separated number every three digits. `NumComma_add_str` is the
#'  function to convert a number in a string to a comma-separated number every
#'   three digits.
#' @param num Numbers (strings) separated by a comma every three digits.
#' @param convert If `TRUE`, will return as numeric. The default value is `FALSE`,
#'  and numeric values are treated as strings.
#' @param warnig.message If TRUE, a warning message is displayed when a
#' non-convertible value is submitted and the input value is returned without
#' conversion; if FALSE, the warning message is hidden and the value of the
#' unconvertible value is returned.
#'
#' @rdname NumComma
#' @importFrom dplyr %>%
#' @return characters or numbers.
#' @examples
#' NumComma_add(2020)
#' NumComma_add(c(2019, 2020, 2021))
#' NumComma_add(c(123456, 1234), digit = 4)
#' NumComma_add(c(123456, 1234), small.num = 2000)
#' NumComma_add_str("This year is 2020 and next year is 2021.")
#' NumComma_add_str("This is 123456, but that's 2020.", digit = 4)
#' NumComma_rm(c("1,234", "1,234.56", "1,23", "japan"), warnig.message = FALSE)
#' NumComma_rm(c("1,234", "1,234.56", "1,23", "japan"), convert = TRUE, warnig.message = FALSE)
#' NumComma_rm_str("This year is 2,020 and next year is 2,021.")
#' NumComma_rm_str("This year is 2,0,2,0 and next year is 2,021.")
#' @export
#'

NumComma_rm <- function(num, convert = FALSE, warnig.message = TRUE){
  purrr::map2(num, convert, function(num, convert){
    if(warnig.message){
      if(!stringr::str_detect(num, pattern = "[:digit:]&,") &&
         !stringr::str_detect(num, pattern = "^[:digit:]"))
        warning("Strings containing characters other than numbers and commas cannot be converted.")
    }

    row_num <- num
    comma_n <- stringr::str_count(num, pattern = "[:digit:],[:digit:]")

    if(comma_n > 0 &&
       !(stringr::str_detect(num, pattern = "[:digit:],[:digit:][:digit:][:digit:][:digit:]")) &&
       !(stringr::str_detect(num, pattern = "[:digit:],[:digit:],")) &&
       !(stringr::str_detect(num, pattern = "[:digit:],[:digit:][:digit:],"))){
      for(i in 1:comma_n){
        comma_num <- stringr::str_extract(num, pattern = "[:digit:],[:digit:][:digit:][:digit:]")
        if(is.na(comma_num)){
          num <- row_num
          break
        }
        convert_num <- stringr::str_replace(comma_num, pattern = ",", replacement = "")
        num <- stringr::str_replace(num, pattern = comma_num, replacement = convert_num)
      }

    }

    if(stringr::str_detect(num, pattern = ",")){
      num <- row_num
      if(warnig.message) warning("this format cannot be converted.")
    }

    if(convert) num <- suppressWarnings(as.numeric(num))

    return(num)
  }) %>% unlist()
}

#' @param str str is a string.
#' @rdname NumComma
#' @export
#'

NumComma_rm_str <- function(str){
  purrr::map_chr(str, function(str){
    str_comma <- stringr::str_split(str, pattern = "[^[:digit:]|,]")[[1]]
    stringr::str_sub(str_comma[stringr::str_sub(str_comma, start = -1) ==  ","], start = -1) <- ""
    str_comma[str_comma == ""] <- NA
    str_comma <- stats::na.omit(str_comma)

    if(!(length(str_comma) == 0)){
      str_num <- NumComma_rm(str_comma, convert = FALSE, warnig.message = FALSE)

      for(i in 1:length(str_num)){
        str <- stringr::str_replace(str, pattern = str_comma[i], replacement = str_num[i])
      }
    }

    return(str)
  })
}

#' @param x x is a number or a string of numbers intended.
#' @param digit Converts the number of digits greater than the number specified
#' by `digit` (including the decimal point) to comma-separated numbers only.
#' @param small.num Convert only numbers greater than the number specified in
#' `small.num` (including decimal points) to comma-separated numbers.
#' @rdname NumComma
#' @export

NumComma_add <- function(x, digit = 3, small.num = 999, warnig.message = TRUE){
  purrr::map_chr(x, function(x, digit, small.num){
    if(suppressWarnings(is.na(as.numeric(x)))){
      if(warnig.message) warning("Strings containing characters other than numbers cannot be converted.")
      as.character(x)
    }
    else if(stringr::str_length(x) > digit &&
            x > small.num &&
            is.numeric(x)) prettyNum(x, big.mark = ",",  scientific = FALSE)
    else if(stringr::str_length(x) > digit &&
            !is.numeric(x) &&
            as.numeric(x) > small.num &&
            stringr::str_length(x) == stringr::str_length(as.numeric(x))) prettyNum(x, big.mark = ",",  scientific = FALSE)
    else as.character(x)
  }, digit, small.num)
}

#' @rdname NumComma
#' @export

NumComma_add_str <- function(str, digit = 3, small.num = 999){
  purrr::map_chr(str, function(str, digit, small.num){
    str_num <- stringr::str_split(str, pattern = "[^[:digit:]|.]")[[1]]
    stringr::str_sub(str_num[stringr::str_sub(str_num, start = -1) ==  "."], start = -1) <- ""
    str_num[str_num == ""] <- NA
    str_num <- stats::na.omit(str_num)

    if(!length(str_num) == 0){
      str_comma <- NumComma_add(as.numeric(str_num), digit = digit, small.num = small.num, warnig.message = FALSE)

      for(i in 1:length(str_comma)){
        str <- stringr::str_replace(str, pattern = str_num[i], replacement = str_comma[i])
      }

    }
    return(str)
  }, digit, small.num)
}
